gobblin-core/src/main/java/org/apache/gobblin/converter/parquet/JsonElementConversionFactory.java

```diff
@@ -0,0 +1,695 @@
+      this.name = fieldName;
+      this.nullable = nullable;
+      this.repeated = repeated;
+    }
+
+    /**
+     * Field name from schema
+     * @return
+     */
+    public String getName() {
+      return this.name;
+    }
+
+    /**
+     * is field nullable
+     * @return
+     */
+    public boolean isNullable() {
+      return this.nullable;
+    }
+
+    /**
+     * Returns a {@link PrimitiveType} schema
+     * @return
+     */
+    protected Type schema() {
+      return new PrimitiveType(repeated ? REPEATED : optionalOrRequired(), getTargetType(), getName());
+    }
+
+    /**
+     * Convert value
+     * @param value is JsonNull will return null if allowed or exception if not allowed
+     * @return Parquet safe type
+     */
+    public Object convert(JsonElement value) {
+      if (value.isJsonNull()) {
+        if (this.nullable) {
+          return null;
+        }
+        throw new RuntimeException("Field: " + getName() + " is not nullable and contains a null value");
+      }
+      return convertField(value);
+    }
+
+    /**
+     * Convert JsonElement to Parquet type
+     * @param value
+     * @return
+     */
+    abstract Object convertField(JsonElement value);
+
+    /**
+     * Parquet data type after conversion
+     * @return
+     */
+    public abstract PrimitiveTypeName getTargetType();
+
+    /**
+     * Returns the source type
+     * @return
+     */
+    public abstract SourceType getSourceType();
+
+    /**
+     * Parquet RepetitionType Optional or Required
+     * @return
+     */
+    protected Type.Repetition optionalOrRequired() {
+      return isNullable() ? OPTIONAL : REQUIRED;
+    }
+
+    /**
+     * Is Parquet type repeated in the {@link Group}
+     * @return
+     */
+    protected boolean isRepeated() {
+      return this.repeated;
+    }
+  }
+
+  /**
+   * For converting Complex data types like ARRAY, ENUM, MAP etc.
+   */
+  public static abstract class ComplexConverter extends JsonElementConverter {
+    private JsonElementConverter elementConverter;
+
+    public ComplexConverter(String fieldName, boolean nullable, boolean repeated) {
+      super(fieldName, nullable, repeated);
+    }
+
+    /**
+     * Set a {@link JsonElementConverter} for the elements present within Complex data types.
+     * @param elementConverter
+     */
+    protected void setElementConverter(JsonElementConverter elementConverter) {
+      this.elementConverter = elementConverter;
+    }
+
+    public JsonElementConverter getElementConverter() {
+      return this.elementConverter;
+    }
+
+    @Override
+    public PrimitiveTypeName getTargetType() {
+      throw new UnsupportedOperationException("Complex types does not support PrimitiveTypeName");
+    }
+  }
+
+  /**
+   * For converting complex types which compose same element type.
+   */
+  public static abstract class ComplexConverterForUniformElementTypes extends ComplexConverter {
```

Comment:
Just like java `Object` type is to any java types, is there a super type that unifies all element types? In this case, a single type of `ComplexConverter` can be used for both cases.

It's just an idea. You can drop it if it doesn't satisfy all the requirements.